### PR TITLE
DFH-369 LikeSearch の CompoundColumn でnullを空文字として連結できるようにする。

### PR DESCRIPTION
**Pull request by h-funaki**

### CODE REVIEW

File: dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java [hunk 1]
```diff
@@ -55,6 +56,7 @@
     protected List<String> _originalWildCardList;
     protected List<SpecifiedColumn> _compoundColumnList;
     protected List<Integer> _compoundColumnSizeList;
+    protected boolean _treatCompoundColumnNullAsEmpty;
```
Comment (jflute) [inline dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java]: change _treatCompoundColumnNullAsEmpty to _treatsCompoundColumnNullAsEmpty
("treat" to "treats")

File: dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java [hunk 2]
```diff
@@ -324,6 +330,15 @@ public boolean hasCompoundColumn() {
         return _compoundColumnList;
     }
 
+    public void setCompoundColumnNullAsEmpty() {
```
Comment (jflute) [inline dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java]: The "set" is special word in Java. (used as getter/setter)
LikeSearchOption does not use "set" for option methods.
It uses e.g. optimize, split, escape, ... instead.

So I want to use "set" at least in this class.
How about treatCompoundColumnNullAsEmpty()?

While, treatsCompoundColumnNullAsEmpty() (for determination) already exists.
treatCompoundColumnNullAsEmpty(), treatsCompoundColumnNullAsEmpty() are very confused.
Internal determination methods in LikeSearchOption are almost isXxx(), e.g. isLikePrefix() isAsOrSplit().
So how about isTreatCompoundColumnNullAsEmpty();

Of course, also isTreat...() is not good name.
But an user method should be more simple than an internal method.

**Comment by jflute** [comment]
Thanks.
Can you change the branch "dbflute:master" to "dbflute:develop"?

The branch "master" is always latest release code in this project.
So please pull request to branch "develop".

File: dbflute-runtime/src/main/java/org/dbflute/cbean/ckey/ConditionKey.java [hunk 3]
```diff
@@ -449,10 +450,19 @@ protected ColumnRealName resolveCompoundColumn(ColumnRealName baseRealName, Cond
...
+            realNameList.add(toColumnRealName("coalesce(" + baseRealName + ",\'\')"));
+            for (SpecifiedColumn specifiedColumn : compoundColumnList) {
+                realNameList.add(toColumnRealName("coalesce(" + doResolveCompoundColumn(option, specifiedColumn) + ",\'\')"));
+            }
+        } else {
+            realNameList.add(baseRealName); // already cipher
+            for (SpecifiedColumn specifiedColumn : compoundColumnList) {
+                ColumnRealName resolvedColumnRealName = doResolveCompoundColumn(option, specifiedColumn);
+                realNameList.add(resolvedColumnRealName);      
+            }
```
Comment (jflute) [inline dbflute-runtime/src/main/java/org/dbflute/cbean/ckey/ConditionKey.java]: I want you to refactor this logic if you can.
Please unify for-statement.

**Comment by h-funaki** [comment]
> Thanks.
> Can you change the branch "dbflute:master" to "dbflute:develop"?
> 
> The branch "master" is always latest release code in this project.
> So please pull request to branch "develop".

i understand
changed "master" branch to "develop" branch.

File: dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java (hunk 2 above)
Comment (h-funaki) [reply dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java]: > The "set" is special word in Java. (used as getter/setter)
> LikeSearchOption does not use "set" for option methods.
> It uses e.g. optimize, split, escape, ... instead.
> 
> So I don't want to use "set" at least in this class.
> How about treatCompoundColumnNullAsEmpty()?
> 
> While, treatsCompoundColumnNullAsEmpty() (for determination) already exists.
> treatCompoundColumnNullAsEmpty(), treatsCompoundColumnNullAsEmpty() are very confused.
> Internal determination methods in LikeSearchOption are almost isXxx(), e.g. isLikePrefix() isAsOrSplit().
> So how about isTreatCompoundColumnNullAsEmpty();
> 
> Of course, also isTreat...() is not good name.
> But an user method should be more simple than an internal method.

Noted with thanks. I understand.
How about "compoundsNullAsEmpty()" as the user method and "isNullCompoundedAsEmpty()" as the internal method ?
Then, I think the user method is more simple and the internal method matches "isXxx()" expression.

Comment (h-funaki) [reply dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java]: and more natural english maybe
Null is compounded as empty? -> maybe good
treats compound column null as empty -> wrong (what is the object?)

Comment (jflute) [reply dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java]: > How about "compoundsNullAsEmpty()" as the user method and 
> "isNullCompoundedAsEmpty()" as the internal method ?
Oh, OK, good. That's new for me.

File: dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java [hunk 4]
```diff
@@ -55,6 +56,7 @@
     protected List<String> _originalWildCardList;
     protected List<SpecifiedColumn> _compoundColumnList;
     protected List<Integer> _compoundColumnSizeList;
+    protected boolean _isNullcompoundedAsEmpty;
```
Comment (jflute) [inline dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java]: _isNullcompoundedAsEmpty to _isNull "C" ompoundedAsEmpty

File: dbflute-runtime/src/main/java/org/dbflute/cbean/ckey/ConditionKey.java (hunk 3 above)
Comment (h-funaki) [reply dbflute-runtime/src/main/java/org/dbflute/cbean/ckey/ConditionKey.java]: > I want you to refactor this logic if you can.
> Please unify for-statement.

modified for-statement.

https://github.com/dbflute/dbflute-core/pull/54/commits/3178d878b1e2f463866a5dabcca86c2172091b4e

File: dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java (hunk 2 above)
Comment (h-funaki) [reply dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java]: > How about "compoundsNullAsEmpty()" as the user method and
> "isNullCompoundedAsEmpty()" as the internal method ?
> Oh, OK, good. That's new for me.

renamed them.
https://github.com/dbflute/dbflute-core/pull/54/commits/3178d878b1e2f463866a5dabcca86c2172091b4e

File: dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java (hunk 4 above)
Comment (jflute) [inline dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java]: Frequently an instance variable name does not have prefix 'is'.
For example, _asOrSplit in LikeSearchOption.
An accessor method has it instead like isAsOrSprlit().
So rename to _nullCompoundedAsEmpty.

Comment (h-funaki) [reply dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java]: > Frequently an instance variable name does not have prefix 'is'.
> For example, _asOrSplit in LikeSearchOption.
> An accessor method has it instead like isAsOrSprlit().
> So rename to _nullCompoundedAsEmpty.

It make sense. I understand very well.
https://github.com/dbflute/dbflute-core/pull/54/commits/af6b6ecf2249a76b9e15475f082f5f691c8024e0

Comment (h-funaki) [reply dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java]: > _isNullcompoundedAsEmpty to _isNull "C" ompoundedAsEmpty

sorry, I made a mistake.

Comment (h-funaki) [reply dbflute-runtime/src/main/java/org/dbflute/cbean/coption/LikeSearchOption.java]: modified to "_nullCompoundedAsEmpty"

https://github.com/dbflute/dbflute-core/commit/af6b6ecf2249a76b9e15475f082f5f691c8024e0

File: dbflute-runtime/src/main/java/org/dbflute/cbean/ckey/ConditionKey.java [hunk 5]
```diff
@@ -449,10 +450,19 @@ protected ColumnRealName resolveCompoundColumn(ColumnRealName baseRealName, Cond
...
+        } else {
+            realNameList.add(baseRealName); // already cipher            
+        }
         for (SpecifiedColumn specifiedColumn : compoundColumnList) {
-            realNameList.add(doResolveCompoundColumn(option, specifiedColumn));
+            if (option.isNullCompoundedAsEmpty()) {
+                realNameList.add(toColumnRealName("coalesce(" + doResolveCompoundColumn(option, specifiedColumn) + ",\'\')"));
+            } else {
+                realNameList.add(doResolveCompoundColumn(option, specifiedColumn));
+            }
```
Comment (jflute) [inline dbflute-runtime/src/main/java/org/dbflute/cbean/ckey/ConditionKey.java]: Thank you.
But maybe you can refactor more like this:
```
realNameList.add(xxxxxx(option, baseRealName));
for (SpecifiedColumn specifiedColumn : compoundColumnList) {
    realNameList.add(xxxxxx(option, doResolveCompoundColumn(option, specifiedColumn))));
}
```

Comment (h-funaki) [reply dbflute-runtime/src/main/java/org/dbflute/cbean/ckey/ConditionKey.java]: > Thank you.
> But maybe you can refactor more like this:
> 
> realNameList.add(xxxxxx(option, baseRealName));
> for (SpecifiedColumn specifiedColumn : compoundColumnList) {
>     realNameList.add(xxxxxx(option, doResolveCompoundColumn(option, specifiedColumn))));
> }

refactored the codes as you suggested
https://github.com/dbflute/dbflute-core/pull/54/commits/37bf3a4b02b2c002b6779dc0b16542f0d8618c4e

File: dbflute-runtime/src/main/java/org/dbflute/cbean/ckey/ConditionKey.java [hunk 6]
```diff
@@ -449,14 +450,21 @@ protected ColumnRealName resolveCompoundColumn(ColumnRealName baseRealName, Cond
...
         for (SpecifiedColumn specifiedColumn : compoundColumnList) {
-            realNameList.add(doResolveCompoundColumn(option, specifiedColumn));
+            realNameList.add(resolveCompoundColumnOption(option, doResolveCompoundColumn(option, specifiedColumn)));
         }
         final OnQueryStringConnector stringConnector = option.getStringConnector();
         final String connected = stringConnector.connect(realNameList.toArray());
         return ColumnRealName.create(null, new ColumnSqlName(connected));
     }
+    
+    protected ColumnRealName resolveCompoundColumnOption(ConditionOption option, ColumnRealName columnRealName) {
```
Comment (jflute) [inline dbflute-runtime/src/main/java/org/dbflute/cbean/ckey/ConditionKey.java]: Thanks.
The two methods are bad balance, 
resolveCompoundColumnOption() and doResolveCompoundColumn().

The doResolveCompoundColumn() is already too abstract name in the now
so rename to doResolveCompoundColumnCipher() that the name has purpose.
(Existing methods can be adjusted when adding new methods)
And keep "already cipher" comment to tell someone why first column is not cipher target.

And a nested method has prefix 'do' like doResolveCompoundColumn()
so rename to doResolveCompoundColumnOption().

Comment (h-funaki) [reply dbflute-runtime/src/main/java/org/dbflute/cbean/ckey/ConditionKey.java]: > Thanks.
> The two methods are bad balance,
> resolveCompoundColumnOption() and doResolveCompoundColumn().
> 
> The doResolveCompoundColumn() is already too abstract name in the now
> so rename to doResolveCompoundColumnCipher() that the name has purpose.
> (Existing methods can be adjusted when adding new methods)
> And keep "already cipher" comment to tell someone why first column is not cipher target.
> 
> And a nested method has prefix 'do' like doResolveCompoundColumn()
> so rename to doResolveCompoundColumnOption().

thank you.

Comment (h-funaki) [reply dbflute-runtime/src/main/java/org/dbflute/cbean/ckey/ConditionKey.java]: renamed as you supposed.

https://github.com/dbflute/dbflute-core/pull/54/commits/a2cdd3a82193ced0da4cbb0b6bf12387f79ca97d

**Comment by jflute** [comment]
Great, thank you!!!

**Pull request merged**